rescript/trim_alignment.py

```diff
@@ -221,10 +221,17 @@ def _trim_alignment(expand_alignment_action,
             'FeatureData[Sequence]', primers_fasta)
 
         # expand the existing alignment by addition of primers
+        # Use 'keeplength' to make sure original alignment length
+        # does not change. Otherwise there is potential for the incorrect
+        # region to be extracted from the alignment. This is important
+        # to make sure alignment positions can easily be referenced
+        # to the original alignment. See the 'q2-alignment' plugin,
+        # specifically the 'mafft-add' test cases for examples.
         alignment_with_primers, = expand_alignment_action(
             alignment=aligned_sequences,
             sequences=primers,
             addfragments=True,
+            keeplength=True,
```

Comment:
hey @mikerobeson do we want to hardcode this param? Should we make it True by default but expose the option to the user? Is there ever a case when they might want to not keep the length? In this case this would only happen if there are indels in the primer region, right? which we probably don't want to preserve.

Reply:
These primer indel cases are explicitly tested in the [mafft-add](https://github.com/qiime2/q2-alignment/blob/dev/q2_alignment/tests/test_mafft.py) tests by toggling that option on/off. I was thinking I did not need to reproduce that test case here. 

The issue is that the final alignment length changes after the primer sequences are added. Often, due to common column gaps in the newly combined alignment being removed. From what I can observe this is regardless if new indels are added within the location of the newly aligned primers. If any of the V3V4, V4, V4V5 primers are used no indels are added to the location where the primers align.

Due to the alignment length change after the primers are added to the alignment, the derived positional values of the primers will differ from the proper positions of the original alignment. Thus, when these derived positional values are used, the actual region extracted is not the region we are after, as they are based on the positions of the new alignment.

I suppose I could add another set of small files based on a few of the actual SILVA  alignment sequences. Then I can check the expected vs observed returned alignment length differences. I'd just have to override  `keeplength=True` option and compare.

Reply:
This issue is the trimming aspect... we are deriving the positional values from the resulting alignment after the primers are added. Then using those values to trim from the original alignment. This won't work in all cases as the resulting alignment length can very likely change. Which is why we need to use `keeplength=True`. I think that toggling this option makes obvious sense for the general `mafft-add` function... but not for our specific use case here: finding alignment position trimming locations.

Reply:
okay now I understand! So any length change messes up the positional trimming. I agree, hardcoding sounds necessary in that case.

And agreed, unit tests are not needed if this is explicitly tested in q2-alignment.

Reply:
I just realized that I can likely fix this by adding an `if` statement to check if primers were used. If primers are used then this [line](https://github.com/bokulich-lab/RESCRIPt/blob/49fc648043af56dfafa41a030ac4927bab4f4a83/rescript/trim_alignment.py#L240) could be something like: 

```
result = _trim_all_sequences(aligned_sequences_with_primers, trim_positions)
```

Then the primer sequences can removed from that output prior to saving. This would enable us toggle the `keeplength` option. For instance if users would like to concatenate their edits to some master alignment and not worry about alignment length changes. 🤔

Reply:
Hi @nbokulich, I've added explicit unit test for the keep length option. I decided to forgoe toggling the `--p-keeplengh` parameter, as it is expected that this be used when trying to find the location of added fragments. In fact, the mafft parameter `--mapout` (currently not accessible within the plugin) enforces the mafft `--keeplength` function if it is not supplied. Thus,, I think we are in keeping with the spirit of the tool when it comes to determining alignment positions. :-)

Reply:
Actually, I'm still working on the test cases... I'll let you know when I am done. :-)

Reply:
Okay... I realized that I should actually run my test cases rather than mock them. Should be read to review now. :-)